apollo-router/src/spec/schema.rs

```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
+                                        .flat_map(|ext| ext.implements_interfaces()),
+                                )
+                                .map(implements_interface)
+                                .collect(),
+                        },
+                    )
+                })
+                .collect();
+
+            let interfaces: HashMap<_, _> = compiler
+                .db
+                .interfaces()
+                .iter()
+                .map(|(name, def)| {
+                    (
+                        name.clone(),
+                        Interface {
+                            name: name.clone(),
+                            fields: def
+                                .fields_definition()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.fields_definition()),
+                                )
+                                .map(object_field)
+                                .collect(),
+                            interfaces: def
+                                .implements_interfaces()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.implements_interfaces()),
+                                )
+                                .map(implements_interface)
+                                .collect(),
+                        },
+                    )
+                })
+                .collect();
+
+            let input_field = |field: &hir::InputValueDefinition| {
+                (
+                    field.name().to_owned(),
+                    (
+                        field.ty().into(),
+                        field.default_value().and_then(parse_hir_value),
+                    ),
+                )
+            };
+            let input_types: HashMap<_, _> = compiler
+                .db
+                .input_objects()
+                .iter()
+                .map(|(name, def)| {
+                    (
+                        name.clone(),
+                        InputObjectType {
+                            name: name.clone(),
+                            fields: def
+                                .input_fields_definition()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.input_fields_definition()),
+                                )
+                                .map(input_field)
+                                .collect(),
+                        },
```

Comment:
I'm wondering if it would be more readable if we used `From` impl for that kind of code ?